### PR TITLE
Add macro to generate accessors for RenderObjects

### PR DESCRIPTION
**Pull request by goderbauer**
A simple macro to generate the common getter/setter pairs often found on RenderObjects to get familiar with the macro API.

### CODE REVIEW

File: flutter_example/lib/main.gen.dart [hunk 1]
```diff
@@ -114,3 +92,42 @@ class SimpleDisposable implements Disposable {
...
+  @override
+  void updateRenderObject(BuildContext context,  MyColoredFillRenderBox renderObject) {
+    renderObject.color = color;
+  }
+}
+
+class MyColoredFillRenderBox extends RenderBox {
+  MyColoredFillRenderBox({required Color color}) : _color = color;
+
+  @RenderAccessors(needsPaint: true) Color _color;
```
Comment (goderbauer) [inline flutter_example/lib/main.gen.dart:116]: I wish this annotation could also somehow add the color param and the initializer to the constructor, but didn't see a way to do that.

File: flutter_example/lib/macros/render_accessors.dart [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  final bool needsPaint;
+  final bool needsLayout;
+  final bool needsSemantics;
+
+  @override
+  void visitFieldDeclaration(
+    FieldDeclaration declaration,
+    ClassDeclarationBuilder builder,
+  ) {
+    // TODO(goderbauer): Check that the annotated field is inside a RenderObject. How?
```
Comment (jakemac53) [inline flutter_example/lib/macros/render_accessors.dart:19]: Interesting point, I think this is something we need to enable. I will work on figuring out something that doesn't break our ordering model.

I think it should be ok to give you the previous phases information for parent declarations, without introducing ordering weirdness.

File: flutter_example/lib/main.gen.dart (hunk 1 above)
Comment (jakemac53) [reply flutter_example/lib/main.gen.dart:116]: That would need to be a separate macro (a class declaration macro most likely) - you can see the data class example in the dart vm examples.

File: flutter_example/lib/macros/render_accessors.dart (hunk 2 above)
Comment (goderbauer) [reply flutter_example/lib/macros/render_accessors.dart:19]: https://github.com/jakemac53/macro_prototype/issues/21

**Review by jakemac53 — APPROVED** [review]
This looks good and some good feedback on the need to type check the surrounding class, thanks!

Do you have any other general feedback about the macro system? For fields in particular, do you think it is acceptable to require that they write a private field, and then you generate the public getters/setters?

**Comment by goderbauer** [comment]
> For fields in particular, do you think it is acceptable to require that they write a private field, and then you generate the public getters/setters?

That seems acceptable to me.

> Do you have any other general feedback about the macro system?

It took me a good amount of time to find a useful macro that I felt was implementable with the system. Before implementing this macro I tried to come up with some kind of a macro that makes using Listenables in stateful widgets less boilerplate-y (that was one of the use cases we looked at in our initial brainstorm, see also example in https://gist.github.com/goderbauer/b82efcb154c3fe1814a0d0b60d277def), but I couldn't come up with anything, unfortunately. (This may just be me not being too familiar with what we can do with macros yet). I often got stuck at wanting to add additional logic to a method based on annotations on a field. Wondering if you've any ideas how to make that use case better with macros.

Once I had a doable use case, implementing it felt pretty straight forward and the macro writing API made sense to me.

**Comment by jakemac53** [comment]
> I often got stuck at wanting to add additional logic to a method based on annotations on a field.

You can "wrap" existing methods with extra code (essentially you can add new code before and after the method which has its own scope). See a similar example at https://github.com/jakemac53/macro_prototype/blob/main/flutter_example/lib/macros/auto_dispose.dart. That supports either adding additional code to an existing `dispose` method or creating a whole new one.

**Comment by goderbauer** [comment]
Oh, the combination of ClassDeclarationMacro, MethodDefinitionMacro in https://github.com/jakemac53/macro_prototype/blob/main/flutter_example/lib/macros/auto_dispose.dart is interesting. What I don't like is that the annotation has to moves from the class to the method if you have an explicit dispose defined. But sounds like that could be fixed with https://github.com/jakemac53/macro_prototype/issues/8. In that case, the annotation could always go on the class.

**Pull request merged**